packages/pigeon/lib/dart_generator.dart

```diff
@@ -200,9 +200,26 @@ final BinaryMessenger$nullTag _binaryMessenger;
         final String returnType =
             _makeGenericTypeArguments(func.returnType, nullTag);
         final String castCall = _makeGenericCastCall(func.returnType, nullTag);
+        const String accessor = 'replyMap[\'${Keys.result}\']';
+        final String unwrapper =
+            func.returnType.isNullable ? '' : unwrapOperator;
         final String returnStatement = func.returnType.isVoid
             ? 'return;'
-            : 'return (replyMap[\'${Keys.result}\'] as $returnType$nullTag)$unwrapOperator$castCall;';
+            : 'return ($accessor as $returnType$nullTag)$unwrapper$castCall;';
+        // On iOS we can return nil from functions to accommodate error
+        // handling.  Returning a nil value and not returning an error is an
+        // exception.
+        final String nullCheck =
+            func.returnType.isNullable | func.returnType.isVoid
+                ? ''
+                : '''
+
+} else if (replyMap['${Keys.result}'] == null) {
```

Comment:
The fact that this is before the `if` line in the generating code was super confusing to me in trying to follow this. I think it would be helpful to make each of the if/else if/else blocks a local variable, declared in order, and then use all three below, so that generated code sequence is in order in this code.

Reply:
(Longer term than this PR: I think the complexity has gotten to the point where we should very strongly consider using a templating system, like we do for the generated plugin registrants in `flutter_tool`. We could likely make it *much* easier to read the to-be-generated code that way.)

Reply:
I swapped everything now so it shows up in the correct location now.  Something like mustache might help.  Its hard to guess the ROI right now though.